proxy/envoy/config.go

```diff
@@ -606,6 +608,73 @@ func buildInboundListeners(mesh *proxyconfig.ProxyMeshConfig, sidecar proxy.Node
 	return listeners, clusters
 }
 
+func appendPortToDomains(domains []string, port int) []string {
+	domainsWithPorts := make([]string, len(domains), 2*len(domains))
+	copy(domainsWithPorts, domains)
+
+	for _, domain := range domains {
+		domainsWithPorts = append(domainsWithPorts, domain + ":" + strconv.Itoa(port))
+	}
+
+	return domainsWithPorts
+}
+
+func buildExternalTrafficVirtualHostOnPort(rule *proxyconfig.EgressRule, mesh *proxyconfig.ProxyMeshConfig, port *model.Port) *VirtualHost {
+	var externalTrafficCluster *Cluster
+
+	protocolSuffix := string(port.Protocol)
+	if rule.UseEgressProxy {
+		externalTrafficCluster = buildOutboundCluster("istio-egress.default-"+protocolSuffix, port, nil)
+		externalTrafficCluster.ServiceName = ""
+		externalTrafficCluster.Type = ClusterTypeStrictDNS
+		externalTrafficCluster.Hosts = []Host{{URL: fmt.Sprintf("tcp://%s", mesh.EgressProxyAddress)}}
+	} else {
+		externalTrafficCluster = buildOriginalDSTCluster("orig-dst-cluster"+protocolSuffix, mesh.ConnectTimeout)
+		if port.Protocol == model.ProtocolHTTPS {
+			externalTrafficCluster.SSLContext = &SSLContextExternal{}
+		}
+	}
+
+	externalTrafficRoute := buildDefaultRoute(externalTrafficCluster)
+
+	return &VirtualHost{
+		Name:    rule.Domains[0], // use the first domain as the name
+		Domains: appendPortToDomains(rule.Domains, port.Port),
+		Routes:  []*HTTPRoute{externalTrafficRoute},
+	}
+}
+
+// buildExternalTrafficVirtualHosts builds virtual hosts from egress rule
+func buildExternalTrafficVirtualHostsOnPort(rule *proxyconfig.EgressRule, mesh *proxyconfig.ProxyMeshConfig,
+	port *model.Port) []*VirtualHost {
+	hosts := make([]*VirtualHost, 0)
+
+	host := buildExternalTrafficVirtualHostOnPort(rule, mesh, port)
+	hosts = append(hosts, host)
+
+	return hosts
+}
+
+func addExternalTrafficVirtualHosts(httpConfigs *HTTPRouteConfigs, egressRules map[string]*proxyconfig.EgressRule,
+	mesh *proxyconfig.ProxyMeshConfig) {
+	for _, rule := range egressRules {
+		if len(rule.Domains) < 1 {
```

Comment:
how is this possible ? We are validating right?

Reply:
Double validation, to be sure and in case something goes wrong after the rule is read from the storage. To prevent the line `Name: rule.Domains[0]. // use the first domain as the name` to crash.